dev-tools/harness.js

```diff
@@ -0,0 +1,229 @@
+		this.input = !!options.input;
+		this.output = !!options.output;
+		this.error = !!options.error;
+	}
+
+	async run() {
+		const battleStream = new BattleStreams.BattleStream();
+		const game = this.runGame(this.format, battleStream);
+		const log = () => battleStream.battle && console.error(`\n${battleStream.battle.inputLog.join('\n')}\n`);
+		if (this.input) return game.finally(log);
+		if (!this.error) return game;
+		return game.catch(err => {
+			log();
+			throw err;
+		});
+	}
+
+	async runGame(format, battleStream) {
+		const streams = BattleStreams.getPlayerStreams(battleStream);
+		const spec = {formatid: format, seed: this.prng.seed};
+		const p1spec = this.getPlayerSpec("Bot 1", this.p1options);
+		const p2spec = this.getPlayerSpec("Bot 2", this.p2options);
+
+		const p1 = new RandomPlayerAI(
+			streams.p1, Object.assign({seed: this.newSeed()}, this.p1options)).start();
+		const p2 = new RandomPlayerAI(
+			streams.p2, Object.assign({seed: this.newSeed()}, this.p2options)).start();
+
+		streams.omniscient.write(`>start ${JSON.stringify(spec)}\n` +
+			`>player p1 ${JSON.stringify(p1spec)}\n` +
+			`>player p2 ${JSON.stringify(p2spec)}`);
+
+		let chunk;
+		while ((chunk = await Promise.race([streams.omniscient.read(), p1, p2]))) {
+			if (this.output) console.log(chunk);
+		}
+	}
+
+	// Same as PRNG#generatedSeed, only deterministic.
+	// NOTE: advances this.prng's seed by 4.
+	newSeed() {
+		return [
+			Math.floor(this.prng.next() * 0x10000),
+			Math.floor(this.prng.next() * 0x10000),
+			Math.floor(this.prng.next() * 0x10000),
+			Math.floor(this.prng.next() * 0x10000),
+		];
+	}
+
+	getPlayerSpec(name, options) {
+		if (options.team) return {name, team: options.team};
+		return {name, seed: this.newSeed()};
+	}
+}
+
+module.exports = Runner;
+
+const FORMATS = [
+	'gen7randombattle', 'gen7randomdoublesbattle', 'gen7battlefactory', 'gen6randombattle', 'gen6battlefactory',
+	'gen5randombattle', 'gen4randombattle', 'gen3randombattle', 'gen2randombattle', 'gen1randombattle',
+];
+
+class MultiRunner {
+	constructor(options) {
+		this.options = Object.assign({}, options);
+
+		this.totalGames = options.totalGames;
+
+		this.prng = (options.prng && !Array.isArray(options.prng)) ?
+			options.prng : new PRNG(options.prng);
+		this.options.prng = this.prng;
+
+		this.format = options.format;
+		this.cycle = !!options.cycle;
+		this.all = !!options.all;
+
+		this.async = !!options.async;
+
+		this.formatIndex = 0;
+		this.numGames = 0;
+	}
+
+	async run() {
+		let games = [];
+		let format, lastFormat;
+		let failures = 0;
+		while ((format = this.getNextFormat())) {
+			if (this.all && lastFormat && format !== lastFormat) {
+				// If we ran async, we need to now wait for each game and determine its status.
+				// NOTE: Promise.all doesn't help us here because it will resolve when the first
+				// rejection occurs, we need to wait for *all* the rejections.
+				if (this.async) for (const game of games) await game;
```

Comment:
Actually, ``await`` has the same issue, since a failure would raise an exception.

Reply:
However, as it is implemented, ``game`` never rejects.

Reply:
Yeah, it never rejecting is why this pattern is used.

Reply:
Right, I forgot I changed my approach around catching and tracking failures such that this pattern is wrong (in case you were curious, at one point these loops had a `try..catch` around the `await` and I didn't use the `catch(...) method`, so they did actually work). When I changed to use `catch` I got rid of the unnecessary `try..catch`, not realizing that these could now just use `Promise.all`.